bigtable/client/row_range_test.cc

```diff
@@ -0,0 +1,115 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "bigtable/client/row_range.h"
+
+#include <gmock/gmock.h>
+
+TEST(RowRangeTest, Equal) {
+  const std::string keys[] = {
+    "",
+    "a",
+    "b",
+    "c",
+  };
+
+  int size = sizeof(keys)/sizeof(*keys);
+
+  // Test non-empty combinations (x,y where x!=y) of the keys above.
+  for (int i = 0; i < size; i++) {
+    for (int j = i+1; j < size; j++) {
+      for (int k = 0; k < size; k++) {
+        for (int l = k+1; l < size; l++) {
+          bigtable::RowRange r1(keys[i], keys[j]);
+          bigtable::RowRange r2(keys[k], keys[l]);
+
+          if ((keys[i] == keys[k]) && (keys[j] == keys[l])) {
+            // Ranges must be the same
+            ASSERT_TRUE(  (r1 == r2)) << ": " << r1 << "/" << r2;
+            ASSERT_TRUE(! (r1 != r2)) << ": " << r1 << "/" << r2;
+          } else {
+            // Ranges must be diffetent
+            ASSERT_TRUE(! (r1 == r2)) << ": " << r1 << "/" << r2;
+            ASSERT_TRUE(  (r1 != r2)) << ": " << r1 << "/" << r2;
+          }
+        }
+      }
+    }
+  }
+  // Empty range.
+  ASSERT_TRUE(bigtable::RowRange("a", "a") == bigtable::RowRange("b", "b"));
+  // Infinite range ! = Empty range.
+  ASSERT_TRUE(bigtable::RowRange("", "") != bigtable::RowRange("b", "b"));
+  // Infinite ranges.
+  ASSERT_TRUE(bigtable::RowRange("", "") == bigtable::RowRange("", ""));
+}
+
+TEST(RowRangeTest, Prefix) {
+  EXPECT_EQ(bigtable::RowRange::prefix_range("foo"),
+            bigtable::RowRange("foo", "fop"));
+  EXPECT_EQ(bigtable::RowRange::prefix_range(""),
+            bigtable::RowRange("", ""));
+  EXPECT_EQ(bigtable::RowRange::prefix_range("foo\377\377"),
+            bigtable::RowRange("foo\377\377", "fop"));
+  EXPECT_EQ(bigtable::RowRange::prefix_range("\377\377"),
+            bigtable::RowRange("\377\377", ""));
```

Comment:
I think the successor from `"\377\377"`, is `"\0\0\1", right?  You are treating the strings as NUL terminated, and they should be treated as vectors of characters for these purposes.

Reply:
From the java and Go API, the next permutation string  following "\377\377" is empty string "" (\0\0). How can we represent leading zeros  (\0) in a string?

Reply:
Yes, `std::string` supports any number of NUL (`'\0'`) characters.  You should think of `std::string` as a vector of `char`, with some convenience functions for handling NUL terminated strings.

Reply:
The trailing characters of a key are \0. So, we will be fine  even if we don't explicitly provide the trailing zeros. That is why this successor code work for  the go and Java API. 

For example, if the prefix string is "\0\0\0\xFF\xFF\xFF", the successor would be "\0\0\1" (we don't have to explicitly set the trailing zeros, as  "\0\0\1\0\0\0"). And the algorithm will take care of the leading "\0" (will  provide test case covering that). Thanks,